core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java

```diff
@@ -1226,4 +1232,325 @@ private long createFileWithSingleBlock(AlluxioURI uri) throws Exception {
+    // Only one renamed dir should exist
+    List<FileInfo> existingDirs =
+        mFileSystemMaster.listStatus(new AlluxioURI("/"), ListStatusOptions.defaults());
+    Assert.assertEquals(1, existingDirs.size());
+    // The directory should contain the file
+    List<FileInfo> dirChildren =
+        mFileSystemMaster.listStatus(new AlluxioURI(existingDirs.get(0).getPath()),
+            ListStatusOptions.defaults());
+    Assert.assertEquals(1, dirChildren.size());
+  }
+
+  /**
+   * Tests renaming files concurrently to the same destination will only succeed once.
+   */
+  @Test
+  public void sameDstConcurrentRename() throws Exception {
+    int numThreads = 10;
+    final AlluxioURI[] srcs = new AlluxioURI[numThreads];
+    final AlluxioURI[] dsts = new AlluxioURI[numThreads];
+    for (int i = 0; i < numThreads; i++) {
+      srcs[i] = new AlluxioURI("/file" + i);
+      mFileSystemMaster.createFile(srcs[i], CreateFileOptions.defaults());
+      dsts[i] = new AlluxioURI("/renamed");
+    }
+
+    int errors = concurrentRename(srcs, dsts);
+
+    // We should get an error for all but 1 rename
+    Assert.assertEquals(numThreads - 1, errors);
+
+    List<FileInfo> files =
+        mFileSystemMaster.listStatus(new AlluxioURI("/"), ListStatusOptions.defaults());
+    int renamedFiles = 0;
+    int originalFiles = 0;
+    for (FileInfo file : files) {
+      if (file.getName().startsWith("renamed")) {
+        renamedFiles++;
+      }
+      if (file.getName().startsWith("file")) {
+        originalFiles++;
+      }
+    }
+    // One renamed file should exist, and 9 original source files
+    Assert.assertEquals(numThreads, files.size());
+    Assert.assertEquals(1, renamedFiles);
+    Assert.assertEquals(numThreads - 1, originalFiles);
+  }
+
+  /**
+   * Tests renaming files concurrently from one directory to another succeeds.
+   */
+  @Test
+  public void twoDirConcurrentRename() throws Exception {
+    int numThreads = 10;
+    final AlluxioURI[] srcs = new AlluxioURI[numThreads];
+    final AlluxioURI[] dsts = new AlluxioURI[numThreads];
+    AlluxioURI dir1 = new AlluxioURI("/dir1");
+    AlluxioURI dir2 = new AlluxioURI("/dir2");
+    mFileSystemMaster.createDirectory(dir1, CreateDirectoryOptions.defaults());
+    mFileSystemMaster.createDirectory(dir2, CreateDirectoryOptions.defaults());
+    for (int i = 0; i < numThreads; i++) {
+      srcs[i] = dir1.join("file" + i);
+      mFileSystemMaster.createFile(srcs[i], CreateFileOptions.defaults());
+      dsts[i] = dir2.join("renamed" + i);
+    }
+
+    int errors = concurrentRename(srcs, dsts);
+
+    // We should get no errors
+    Assert.assertEquals(0, errors);
+
+    List<FileInfo> dir1Files =
+        mFileSystemMaster.listStatus(dir1, ListStatusOptions.defaults());
+    List<FileInfo> dir2Files =
+        mFileSystemMaster.listStatus(dir2, ListStatusOptions.defaults());
+
+    Assert.assertEquals(0, dir1Files.size());
+    Assert.assertEquals(numThreads, dir2Files.size());
```

Comment:
not only the file size, but can we also confirm the file names are correct (e.g., they really become `rename{1,2,3...}`?